File: wpilibc/src/test/native/cpp/kinematics/DifferentialDriveOdometryTest.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2019 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include <wpi/math>

#include "frc/kinematics/DifferentialDriveKinematics.h"
#include "frc/kinematics/DifferentialDriveOdometry.h"
#include "gtest/gtest.h"

static constexpr double kEpsilon = 1E-9;

using namespace frc;

TEST(DifferentialDriveOdometry, OneIteration) {
  DifferentialDriveKinematics kinematics{0.381_m};
  DifferentialDriveOdometry odometry{kinematics};

  odometry.ResetPosition(Pose2d());
  const auto& pose = odometry.Update(0.02_m, 0.02_m, Rotation2d());

  EXPECT_NEAR(pose.Translation().X().to<double>(), 0.02, kEpsilon);
  EXPECT_NEAR(pose.Translation().Y().to<double>(), 0.0, kEpsilon);
  EXPECT_NEAR(pose.Rotation().Radians().to<double>(), 0.0, kEpsilon);
}

TEST(DifferentialDriveOdometry, QuarterCircle) {
  DifferentialDriveKinematics kinematics{0.381_m};
  DifferentialDriveOdometry odometry{kinematics};

  odometry.ResetPosition(Pose2d());
  const auto& pose = odometry.Update(0.0_m, units::meter_t(5 * wpi::math::pi),
                                     Rotation2d(90.0_deg));

  EXPECT_NEAR(pose.Translation().X().to<double>(), 5.0, kEpsilon);
  EXPECT_NEAR(pose.Translation().Y().to<double>(), 5.0, kEpsilon);
  EXPECT_NEAR(pose.Rotation().Degrees().to<double>(), 90.0, kEpsilon);
}

File: wpilibj/src/test/java/edu/wpi/first/wpilibj/kinematics/DifferentialDriveOdometryTest.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2019 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package edu.wpi.first.wpilibj.kinematics;

import org.junit.jupiter.api.Test;

import edu.wpi.first.wpilibj.geometry.Pose2d;
import edu.wpi.first.wpilibj.geometry.Rotation2d;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;

class DifferentialDriveOdometryTest {
  private static final double kEpsilon = 1E-9;
  private final DifferentialDriveKinematics m_kinematics = new DifferentialDriveKinematics(0.381);
  private final DifferentialDriveOdometry m_odometry = new DifferentialDriveOdometry(m_kinematics);

  @Test
  void testOneIteration() {
    m_odometry.resetPosition(new Pose2d());
    var pose = m_odometry.update(0.02, 0.02, Rotation2d.fromDegrees(0));

    assertAll(
        () -> assertEquals(0.02, pose.getTranslation().getX(), kEpsilon),
        () -> assertEquals(0.00, pose.getTranslation().getY(), kEpsilon),
        () -> assertEquals(0.00, pose.getRotation().getRadians(), kEpsilon)
    );
  }

  @Test
  void testQuarterCircle() {
    m_odometry.resetPosition(new Pose2d());
    var pose = m_odometry.update(0.0, 5 * Math.PI, Rotation2d.fromDegrees(90.0));

    assertAll(
        () -> assertEquals(pose.getTranslation().getX(), 5.0, kEpsilon),
        () -> assertEquals(pose.getTranslation().getY(), 5.0, kEpsilon),
        () -> assertEquals(pose.getRotation().getDegrees(), 90.0, kEpsilon)
    );
  }
}
